Issue 131: increase adaptability

al-niessner opened the pull request:
## 🗒️ Summary
Changed endpoints to be more adaptable to changes

## ⚙️ Test Data and/or Report
Ran tests developed previously to check the basic combinations as step before integration testing:

```
$ service/ut/regress.sh 
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-4.6.9, py-1.8.1, pluggy-0.13.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/niessner/Projects/PDS/registry-api
collected 24 items                                                             

service/ut/base_ut.py::test_bad_group PASSED                             [  4%]
service/ut/base_ut.py::test_bad_lidvid PASSED                            [  8%]
service/ut/base_ut.py::TestBundles::test_bundles PASSED                  [ 12%]
service/ut/base_ut.py::TestBundles::test_collections PASSED              [ 16%]
service/ut/base_ut.py::TestBundles::test_collections_all PASSED          [ 20%]
service/ut/base_ut.py::TestBundles::test_collections_latest PASSED       [ 25%]
service/ut/base_ut.py::TestBundles::test_lidvid PASSED                   [ 29%]
service/ut/base_ut.py::TestBundles::test_lidvid_all PASSED               [ 33%]
service/ut/base_ut.py::TestBundles::test_lidvid_latest PASSED            [ 37%]
service/ut/base_ut.py::TestBundles::test_products PASSED                 [ 41%]
service/ut/base_ut.py::TestCollections::test_bundles PASSED              [ 45%]
service/ut/base_ut.py::TestCollections::test_collections PASSED          [ 50%]
service/ut/base_ut.py::TestCollections::test_lidvid PASSED               [ 54%]
service/ut/base_ut.py::TestCollections::test_lidvid_all PASSED           [ 58%]
service/ut/base_ut.py::TestCollections::test_lidvid_latest PASSED        [ 62%]
service/ut/base_ut.py::TestCollections::test_products PASSED             [ 66%]
service/ut/base_ut.py::TestCollections::test_products_all PASSED         [ 70%]
service/ut/base_ut.py::TestCollections::test_products_latest PASSED      [ 75%]
service/ut/base_ut.py::TestProducts::test_bundles PASSED                 [ 79%]
service/ut/base_ut.py::TestProducts::test_collections PASSED             [ 83%]
service/ut/base_ut.py::TestProducts::test_lidvid PASSED                  [ 87%]
service/ut/base_ut.py::TestProducts::test_lidvid_all PASSED              [ 91%]
service/ut/base_ut.py::TestProducts::test_lidvid_latest PASSED           [ 95%]
service/ut/base_ut.py::TestProducts::test_products PASSED                [100%]

========================== 24 passed in 1.09 seconds =======================
```


## ♻️ Related Issues
#141 
#133
#131 
#118 
#12
#11

Commit:
started the bulk of the rewrite

Commit:
more refactoring

Commit:
more refactoring

Commit:
last of major refactoring hopefully

Commit:
doc update

Commit:
doc update

Commit:
doc update

Commit:
doc update

Commit:
children are working

Commit:
code moved and now compiler error free

Commit:
trying to find endpoints

Commit:
merged main

Commit:
turns out swagger is not happy

Having /{group} and /{id} is too much for swagger as it is not smart enough at the Java layer to differentiate. Had to put gid (group ID) and uid (unique ID as in lid, lidvid, or doi or...) in the front to separate them out.

Commit:
back in business

/gid/{group} and /uid/{identifier} are working again. Need to test the parent/child relationships but seems to be on the right track.

/ still is broken but it seems to be a spring 2.6.6 thing. Not the only person in the world to have this problem but it is not about finding / as much as finding the redirect to /swagger-ui.html. Need to track that down.

Commit:
API bug fixes

Commit:
fix lidvid lookup truncation

Commit:
fix lid to all lidvids problem

Commit:
need to populate the array to sort it

al-niessner commented:
@jordanpadams @tloubrieu-jpl 

Took me longer than expected to weed out the last of the errors but all tests work again. Should pass integration tests now too.

Had to add gid and uid to URL because java generated code from swagger could not handle /{group}.

**old** -> **new** mapping

/bundles -> /gid/bundle
/collections -> /gid/collection
/products -> /any
*products but not bundle or collection* -> /product

/{bundles,collections,products}/{id} -> /uid/{id}
/{bundles,collections,products}/{id}/latest -> /uid/{id}/latest
/{bundles,collections,products}/{id}/all -> /uid/{id}/all

/bundles/{id}/collections -> /uid/{id}/referencing/collection **children**
/bundles/{id}/products -> /uid/{id}/referencing/{any,product} **grandchildren**
/collections/{id}/products -> /uid/{id}/referencing/{any,product} **children**
/products/{id}/bundles -> /uid/{id}/referencing/bundle **grandparents**

I think you can fill in the rest for the various directions. The thing to note is that the relationship is determined from looking up the product class for the id and the group given.

Right now any == product. I need to change what was called the "preset" conditions for filtering product types to make product not be collection or bundle. Thought it would be best under a new ticket.

PS:

can also reverse the referencing:

/bundles/{id}/collection == /uid/{id}/referencing/collection == /gid/collection/referencing/{id}

given the id is a bundle. I made the structure bi-directional in meaning.

These are not bi directional in meaning

/uid/{image id}/referencing/document != /gid/document/referencing/{image id}

/uid/{image id}/referencing/document means all of the product class document that are directly referenced by {image id}.

 /gid/document/referencing/{image id} means all of the product class document that directly reference the {image id}

Only grandparents get a hop.

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/ReferencingLogic.java:
```diff
@@ -0,0 +1,36 @@
+	/**
+	 * Find all of the PDS items of the given Group that reference the specified ID.
+	 * 
+	 * @param input
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/ReferencingLogic.java:
```diff
@@ -0,0 +1,36 @@
+{
+	/**
+	 * Map the set of PDS constraints that define just PDS items that make up this Group. 
+	 * @return
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/model/RefLogicAny.java:
```diff
@@ -0,0 +1,75 @@
+			             "product_class"));
+	}
+
+	private RequestAndResponseContext search(ControlContext context, UserContext input, boolean isIdToGroup)
```
ignore

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/controller/SwaggerJavaTransmuter.java:
```diff
@@ -0,0 +1,255 @@
+
+	private boolean proxyRunsOnDefaultPort()
+	{
+        return (((this.context.getScheme() == "https")  && (this.context.getServerPort() == 443)) 
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/model/ReferencingLogicTransmuter.java:
```diff
@@ -0,0 +1,70 @@
+	Collection(new RefLogicCollection(), "Product_Collection", "collection"),
+	Product(new RefLogicProduct(), "", "product");
+	
+	final private ReferencingLogic refLogic;
```
ignore

al-niessner replied:
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/ReferencingLogic.java:
```diff
@@ -0,0 +1,36 @@
+	 * Find all of the PDS items of the given Group that reference the specified ID.
+	 * 
+	 * @param input
+	 * @return
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/ReferencingLogic.java:
```diff
@@ -0,0 +1,36 @@
+	/**
+	 * Find all of the PDS items of the given ID that reference the specified Group.
+	 * 
+	 * @param input
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/controller/SwaggerJavaTransmuter.java:
```diff
@@ -0,0 +1,255 @@
+	private boolean proxyRunsOnDefaultPort()
+	{
+        return (((this.context.getScheme() == "https")  && (this.context.getServerPort() == 443)) 
+                || ((this.context.getScheme() == "http")  && (this.context.getServerPort() == 80)));
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/ReferencingLogic.java:
```diff
@@ -0,0 +1,36 @@
+	 * Find all of the PDS items of the given ID that reference the specified Group.
+	 * 
+	 * @param input
+	 * @return
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/controller/SwaggerJavaTransmuter.java:
```diff
@@ -0,0 +1,255 @@
+
+	private boolean proxyRunsOnDefaultPort()
+	{
+        return (((this.context.getScheme() == "https")  && (this.context.getServerPort() == 443)) 
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/controller/SwaggerJavaTransmuter.java:
```diff
@@ -0,0 +1,255 @@
+	private boolean proxyRunsOnDefaultPort()
+	{
+        return (((this.context.getScheme() == "https")  && (this.context.getServerPort() == 443)) 
+                || ((this.context.getScheme() == "http")  && (this.context.getServerPort() == 80)));
```
fixed

Commit:
sonatype fixes

jordanpadams commented:
@al-niessner

I think we may need to figure out something other than `uid`/`gid`. Not super intuitive. what if we made that `products` (since everything in the PDS is a product). what if we really generalized this even further and remove any notion of "collection"/"bundle" from the API terminology

for below, let's say `{id}` is a PDS identifier, and `{groupid}` which is mostly == `class` is this idea of some subset of data based upon some set of product classes or other registry groupings (e.g. bundles/collections/observational/documentation/"data")

**/product-classes/{groupid}**

- /bundles -> /gid/bundle -> **/product-classes/bundles**
- /collections -> /gid/collection -> **/product-classes/collections**
- /products -> /any -> **/products**


**/products/{id}**

- /{bundles,collections,products}/{id} -> /uid/{id} -> **/products/{id}**
- /{bundles,collections,products}/{id}/latest -> /uid/{id}/latest -**> /products/{id}/latest**
- /{bundles,collections,products}/{id}/all -> /uid/{id}/all -> **/products/{id}/all**

For this section, there are actually 2 ways you can "reference" something in the PDS:
1. as parent/child relationship in terms of the data set hierarchy
2. by reference to the LID for documentation / traceability purposes (e.g. context products are referenced by LID/LIDVID, source products for derived products, etc.)

So what do we think about:

- **/products/{id}/references/{groupid}**
- /bundles/{id}/collections -> /uid/{id}/referencing/collection -> **/products/{id}/references/collections**
- /bundles/{id}/products -> /uid/{id}/referencing/{any,product} -> **/products/{id}/references/products** - @al-niessner so for this particular case, how would we know which references are the child products of this bundle versus those that are just referenced by the bundle metadata? also, I'm guessing in this case, we are assuming an implicit "family" relationship between bundles and their grandchildren?
- /collections/{id}/products -> /uid/{id}/referencing/{any,product} - **/products/{id}/references/products** - similar question/comment to above
- /products/{id}/bundles -> /uid/{id}/referencing/bundle -> **/products/{id}/referenced_by/bundles**
- /products/{id}/referencing/observational -> **/products/{id}/references/observational** - all references in the product
- /products/{id}/referencing/observational -> **/products/{id}/referenced_by/observational** - all products that reference this product


thoughts?

Commit:
Merge branch 'main' of github.com:NASA-PDS/registry-api into issue_131

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/model/RefLogicProduct.java:
```diff
@@ -0,0 +1,64 @@
+import gov.nasa.pds.api.registry.search.RequestConstructionContextFactory;
+import gov.nasa.pds.api.registry.search.SearchRequestFactory;
+
+class RefLogicProduct extends RefLogicAny implements ReferencingLogic
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/controller/SwaggerJavaTransmuter.java:
```diff
@@ -0,0 +1,255 @@
+	private boolean proxyRunsOnDefaultPort()
+	{
+        return ((("https".equals(this.context.getScheme()))  && (this.context.getServerPort() == 443)) 
+                || (("http".equals(this.context.getScheme()))  && (this.context.getServerPort() == 80)));
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/controller/SwaggerJavaTransmuter.java:
```diff
@@ -0,0 +1,255 @@
+
+	private boolean proxyRunsOnDefaultPort()
+	{
+        return ((("https".equals(this.context.getScheme()))  && (this.context.getServerPort() == 443)) 
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/model/RefLogicCollection.java:
```diff
@@ -0,0 +1,89 @@
+import gov.nasa.pds.api.registry.search.RequestConstructionContextFactory;
+import gov.nasa.pds.api.registry.search.SearchRequestFactory;
+
+class RefLogicCollection extends RefLogicAny implements ReferencingLogic
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/model/RefLogicBundle.java:
```diff
@@ -17,30 +17,50 @@
  * @author karpenko
  */
-public class BundleDAO
+class RefLogicBundle extends RefLogicAny implements ReferencingLogic
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/model/RefLogicAny.java:
```diff
@@ -0,0 +1,75 @@
+import gov.nasa.pds.api.registry.exceptions.UnknownGroupNameException;
+import gov.nasa.pds.api.registry.search.QuickSearch;
+
+class RefLogicAny implements ReferencingLogic
```
fixed

al-niessner commented:
@jordanpadams @tloubrieu-jpl 

I dislike /products for 2 reasons

1. /products or /gid or /uid is a completely artificial artifact from swagger-java
2. products is way too overloaded already and using it again just adds to confusion and misinterpretation

I would lead with /a and /b to show it is simply an artifact of the tool. Use /p over /products if that makes you happy and /g over /groups for same reason that groups/class/classes are already way overloaded.

As much as I hate to state it, references and referenced_by are clearer. Can we shorten it to ref and ref_by? Less typing is better.

As stated in meeting but will document here, limitations in swagger-java mean that /{group} and /{id} must have different prefixes like /gid/{group} and /uid/{id} but you can make gid and uid anything you want as long as they are different. Eg, /g and /p or /c and /p but not /p and /p.

Commit:
sonatype fixes

jordanpadams commented:
@al-niessner we should probably talk about this again on Thursday.

> /p over /products if that makes you happy and /g over /groups

This is a no-go. I know we all like less typing, but we need words (preferably plural where possible) to be as RESTful as possible, as well as self-documenting. Same goes for refs/ref_by. 

Per the issue with products being overloaded, I am looking to remove that overloading by making `any` == `products`. We remove this idea that a `product` is a "data product", it is anything. You can then narrow from there to `observational`/`document`/etc.

Additionally, per my question here, in this particular case`/products/{BUNDLE_ID}/references/products`, how can we ensure we get just the grandchildren of a bundle? no other references to the bundle, just "familial" references?

al-niessner commented:
> @al-niessner we should probably talk about this again on Thursday.
> 
> Per the issue with products being overloaded, I am looking to remove that overloading by making `any` == `products`. We remove this idea that a `product` is a "data product", it is anything. You can then narrow from there to `observational`/`document`/etc.
> 
> Additionally, per my question here, in this particular case`/products/{BUNDLE_ID}/references/products`, how can we ensure we get just the grandchildren of a bundle? no other references to the bundle, just "familial" references?

You have a group naming problem is all. You can have these definitions:

- any means no restrictions
- bundle means product class is Product Bundle
- collection means product class is Product Collection
- product means product class is **NOT** Product Bundle **NOR** Product Collection

It would then give you:

- /products/{bundle id}/references/any --> direct references + children + grandchildren
- /products/{bundle id}/references/collection -> children
- /products/{bundle id}/references/product -> grandchildren

You can have any group name you want and it can mean anything you want EXCEPT it must not require context like it means product sometimes and any other times. Group names should be singular too (does not break restful) because it is a noun that represents a plural (group). Also, you will want a group that is everything or anything as well.

Once the names and definitions are chosen, can update to match.

Commit:
builds again

Commit:
test validate work is correct

Commit:
sonatype fixes

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/util/GroupConstraintImpl.java:
```diff
@@ -0,0 +1,45 @@
+
+import gov.nasa.pds.api.registry.GroupConstraint;
+
+public class GroupConstraintImpl implements GroupConstraint
```
fixed

Commit:
sonatype fixes

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/model/RefLogicBundle.java:
```diff
@@ -0,0 +1,184 @@
+    
+    /**
+     * Get all versions of bundle's collections by bundle LIDVID.
+     * @param bundleLidVid bundle LIDVID (Could not pass LID here)
```
fixed

Commit:
try for immutability

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/util/GroupConstraintImpl.java:
```diff
@@ -0,0 +1,48 @@
+public class GroupConstraintImpl implements GroupConstraint
+{
+	final private Map<String,List<String>> all;
+	final private Map<String,List<String>> any;
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/util/GroupConstraintImpl.java:
```diff
@@ -0,0 +1,48 @@
+{
+	final private Map<String,List<String>> all;
+	final private Map<String,List<String>> any;
+	final private Map<String,List<String>> not;
```
fixed

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/util/GroupConstraintImpl.java:
```diff
@@ -0,0 +1,48 @@
+@Immutable
+public class GroupConstraintImpl implements GroupConstraint
+{
+	final private Map<String,List<String>> all;
```
fixed

tloubrieu-jpl commented:
Hi @al-niessner , @jordanpadams 

I am not finding the ticket where Al's full proposal for refactoring the URL path is (although I have pasted your proposal in a note of mine).
Sorry if it feels like I am ignoring some aspects of Al's proposal but I am trying to stick to some design principles that I am providing here and the ticket objective (and Al's proposal) was mainly focused on simplification of the registry API code which is important but not our main driver I think.

Here is my proposal, that we can discuss this afternoon:

**Drivers for the design:**
In this order of priority:
1. Transparently shows the PDS4 model (as much as possible) so not to re-invent a PDS model and require to document new concepts specifically for the API (for example use relationships as specified in https://pds.nasa.gov/datastandards/documents/dd/current/PDS4_PDS_DD_1H00.html#N-2021801474)
2. Follow the web restful API design principle and best practices (for example resources specified as nouns are preferred over verbs in the URL path)
3. Refer to use cases to validate the design
4. Have in mind a simple implementation which will help with the maintainability of the code and the readability of the behavior of the API


**Search and identifier resolution:**

For all products:
http://pds.nasa.gov/api/search/1.0/products/{id}

For specific groups of products (classes):
[http://pds.nasa.gov/api/search/1.0/](http://pds.nasa.gov/api/search/1.0/products/){class} as found in https://pds.nasa.gov/datastandards/documents/dd/current/PDS4_PDS_DD_1H00.html#d5e85, by taking the part after Product_ and making it lower case and plural, for example: 
- [http://pds.nasa.gov/api/search/1.0/](http://pds.nasa.gov/api/search/1.0/products/)softwares
- [http://pds.nasa.gov/api/search/1.0/](http://pds.nasa.gov/api/search/1.0/products/)bundles 
- [http://pds.nasa.gov/api/search/1.0/](http://pds.nasa.gov/api/search/1.0/products/)observationals

_Note_ the list given in  https://pds.nasa.gov/datastandards/documents/dd/current/PDS4_PDS_DD_1H00.html#d5e85 is sometimes not consistent in having nouns as the part after  Product_ although most of them are nouns. Rest-ful apis prefers nouns (see https://restfulapi.net/resource-naming/ ). We could replace the adjective or verbs by related nouns:
- browse --> browseable (anything better ?)
- observational --> observation 

in swagger specification we would only have this endpoint entry:
[http://pds.nasa.gov/api/search/1.0/](http://pds.nasa.gov/api/search/1.0/products/){class}  where class is 'products' (anything) or other specific classes (softwares, bundles...)

**Referencing:**
Post used a reference: https://www.moesif.com/blog/technical/api-design/REST-API-Design-Best-Practices-for-Sub-and-Nested-Resources/

From any identifier resolution URL (for example http://pds.nasa.gov/api/search/1.0/observationals/{id} or [http://pds.nasa.gov/api/search/1.0/](http://pds.nasa.gov/api/search/1.0/products/)bundles /{id}) we can get their references by roles of the association as shown on the diagram on page https://pds.nasa.gov/datastandards/documents/im/current/index_1H00.html

_For example:_
- bundles has-member collections --> bundles/{id}/member-collections returns list of collection ids
- collections has-member observationals --> collections/{id}/member-observationals returns list of observationals and documents
- collections as-member documents --> collections/{id}/document-members
- bundles has members collections have members observationals or documents --> /bundles/{id}/member-collections/member-observationals (but should we do that or rely on the user to do 2 requests instead of one, since I don't know the exact use case, I would lean toward the later) 
- observationals is-member-of collections --> observationals/{id}/owner-collections returns list of collection ids
- documents is-member-of collections -->  documents/{id}/owner-collections returns list of collection ids
- collections is-member-of bundles--> collections/{id}/owner-bundles returns list of bundles id
- observationals is-member-of collections is-member-of bundles --> observationals/{id}/owner-collections/owner-bundles (but should we do that or rely on the user to do 2 requests instead of one) 

_Alternate option_ is not to have instead of collection-members only members or instead of bundle-owners only owners . That would less explicit for the user but more simple and that might be obvious enough/documented elsewhere for the users.

products/{id} would not have any referencing besides the generic referencing that a product might have, could be for example Citation_Information which applies to any  class of product (see https://pds.nasa.gov/datastandards/documents/dd/current/PDS4_PDS_DD_1H00.html#N-2006280352). Although we don't have specific entries with identifiers for dois in the registry yet.

al-niessner commented:
/products/{id}/[containers,members]/[all,latest]
/products/{id}/[references,referenced_by}]{group name}/[all,latest]
/products/{id}/[all,latest]
/groups/{group name}/[all,latest]

tloubrieu-jpl commented:
CANCELLED

tloubrieu-jpl commented:
@al-niessner I am seeing that you developed regression tests in python. We would like to consolidate the postman test suite as well (which can be run also in command line), could you work with that in the future ? Thanks

If we can have unit tests they would be developed in Java but I believe what you did is integration tests since they require the registry database to be up and populated.

Let me know if you have questions on that.

Thanks

al-niessner commented:
@tloubrieu-jpl 

Yes, will move to postman when postman checks are integrated into checks done a pull request.

Commit:
merge from main

tloubrieu-jpl commented:
Hi @al-niessner I am getting an error when I do the request ``http://localhost:8081/` which redirects on `http://localhost:8081/swagger-ui.html`

Error sent is:
<img width="601" alt="image" src="https://user-images.githubusercontent.com/60993872/173134215-33593b45-2530-47af-bab6-04cca7287258.png">


Logs give:
```
2022-06-10 21:08:53.853 DEBUG 75288 --- [nio-8081-exec-1] o.a.tomcat.util.net.SocketWrapperBase    : Socket: [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]], Read from buffer: [0]
2022-06-10 21:08:53.855 DEBUG 75288 --- [nio-8081-exec-1] org.apache.tomcat.util.net.NioEndpoint   : Socket: [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]], Read direct from socket: [920]
2022-06-10 21:08:53.860 DEBUG 75288 --- [nio-8081-exec-1] o.a.coyote.http11.Http11InputBuffer      : Received [GET / HTTP/1.1
Host: localhost:8081
Connection: keep-alive
sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="102", "Google Chrome";v="102"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "macOS"
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
Sec-Fetch-Site: none
Sec-Fetch-Mode: navigate
Sec-Fetch-User: ?1
Sec-Fetch-Dest: document
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cookie: _xsrf=2|7ae24c7b|1c43e20182b08315fc49ef578046eb03|1653281851; username-localhost-8888="2|1:0|10:1653456473|23:username-localhost-8888|44:YzM1ZTA5YTUzYTQzNDVlZjhkODUzZjkwZjRiYTBkMzE=|5bd02739555131d2cd77b80629b63f111fae2a623c2f758727c004f291eb729b"

]
2022-06-10 21:08:53.875 DEBUG 75288 --- [nio-8081-exec-1] o.a.t.util.http.Rfc6265CookieProcessor   : Cookies: Parsing b[]: _xsrf=2|7ae24c7b|1c43e20182b08315fc49ef578046eb03|1653281851; username-localhost-8888="2|1:0|10:1653456473|23:username-localhost-8888|44:YzM1ZTA5YTUzYTQzNDVlZjhkODUzZjkwZjRiYTBkMzE=|5bd02739555131d2cd77b80629b63f111fae2a623c2f758727c004f291eb729b"
2022-06-10 21:08:53.885 DEBUG 75288 --- [nio-8081-exec-1] o.a.c.authenticator.AuthenticatorBase    : Security checking request GET /
2022-06-10 21:08:53.888 DEBUG 75288 --- [nio-8081-exec-1] org.apache.catalina.realm.RealmBase      :   No applicable constraints defined
2022-06-10 21:08:53.889 DEBUG 75288 --- [nio-8081-exec-1] o.a.c.authenticator.AuthenticatorBase    : Not subject to any constraint
2022-06-10 21:08:53.906 DEBUG 75288 --- [nio-8081-exec-1] org.apache.tomcat.util.http.Parameters   : Set encoding to UTF-8
2022-06-10 21:08:53.918 DEBUG 75288 --- [nio-8081-exec-1] o.s.web.servlet.DispatcherServlet        : GET "/", parameters={}
2022-06-10 21:08:53.926 DEBUG 75288 --- [nio-8081-exec-1] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to gov.nasa.pds.api.registry.configuration.ApiDocumentation#index()
/swagger-ui.html
2022-06-10 21:08:53.944 DEBUG 75288 --- [nio-8081-exec-1] o.s.web.servlet.view.RedirectView        : View name 'redirect:', model {}
2022-06-10 21:08:53.959 DEBUG 75288 --- [nio-8081-exec-1] o.s.web.servlet.DispatcherServlet        : Completed 302 FOUND
2022-06-10 21:08:53.966 DEBUG 75288 --- [nio-8081-exec-1] o.a.coyote.http11.Http11InputBuffer      : Before fill(): parsingHeader: [true], parsingRequestLine: [true], parsingRequestLinePhase: [0], parsingRequestLineStart: [0], byteBuffer.position(): [0], byteBuffer.limit(): [0], end: [920]
2022-06-10 21:08:53.971 DEBUG 75288 --- [nio-8081-exec-1] o.a.tomcat.util.net.SocketWrapperBase    : Socket: [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]], Read from buffer: [0]
2022-06-10 21:08:53.973 DEBUG 75288 --- [nio-8081-exec-1] org.apache.tomcat.util.net.NioEndpoint   : Socket: [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]], Read direct from socket: [0]
2022-06-10 21:08:53.976 DEBUG 75288 --- [nio-8081-exec-1] o.a.coyote.http11.Http11InputBuffer      : Received []
2022-06-10 21:08:53.985 DEBUG 75288 --- [nio-8081-exec-1] o.apache.coyote.http11.Http11Processor   : Socket: [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]], Status in: [OPEN_READ], State out: [OPEN]
2022-06-10 21:08:53.990 DEBUG 75288 --- [nio-8081-exec-1] org.apache.tomcat.util.net.NioEndpoint   : Registered read interest for [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]]
2022-06-10 21:08:54.006 DEBUG 75288 --- [nio-8081-exec-3] o.a.coyote.http11.Http11InputBuffer      : Before fill(): parsingHeader: [true], parsingRequestLine: [true], parsingRequestLinePhase: [0], parsingRequestLineStart: [0], byteBuffer.position(): [0], byteBuffer.limit(): [0], end: [920]
2022-06-10 21:08:54.011 DEBUG 75288 --- [nio-8081-exec-3] o.a.tomcat.util.net.SocketWrapperBase    : Socket: [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]], Read from buffer: [0]
2022-06-10 21:08:54.018 DEBUG 75288 --- [nio-8081-exec-3] org.apache.tomcat.util.net.NioEndpoint   : Socket: [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]], Read direct from socket: [935]
2022-06-10 21:08:54.019 DEBUG 75288 --- [nio-8081-exec-3] o.a.coyote.http11.Http11InputBuffer      : Received [GET /swagger-ui.html HTTP/1.1
Host: localhost:8081
Connection: keep-alive
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
Sec-Fetch-Site: none
Sec-Fetch-Mode: navigate
Sec-Fetch-User: ?1
Sec-Fetch-Dest: document
sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="102", "Google Chrome";v="102"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "macOS"
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cookie: _xsrf=2|7ae24c7b|1c43e20182b08315fc49ef578046eb03|1653281851; username-localhost-8888="2|1:0|10:1653456473|23:username-localhost-8888|44:YzM1ZTA5YTUzYTQzNDVlZjhkODUzZjkwZjRiYTBkMzE=|5bd02739555131d2cd77b80629b63f111fae2a623c2f758727c004f291eb729b"

]
2022-06-10 21:08:54.021 DEBUG 75288 --- [nio-8081-exec-3] o.a.t.util.http.Rfc6265CookieProcessor   : Cookies: Parsing b[]: _xsrf=2|7ae24c7b|1c43e20182b08315fc49ef578046eb03|1653281851; username-localhost-8888="2|1:0|10:1653456473|23:username-localhost-8888|44:YzM1ZTA5YTUzYTQzNDVlZjhkODUzZjkwZjRiYTBkMzE=|5bd02739555131d2cd77b80629b63f111fae2a623c2f758727c004f291eb729b"
2022-06-10 21:08:54.023 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.authenticator.AuthenticatorBase    : Security checking request GET /swagger-ui.html
2022-06-10 21:08:54.023 DEBUG 75288 --- [nio-8081-exec-3] org.apache.catalina.realm.RealmBase      :   No applicable constraints defined
2022-06-10 21:08:54.024 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.authenticator.AuthenticatorBase    : Not subject to any constraint
2022-06-10 21:08:54.025 DEBUG 75288 --- [nio-8081-exec-3] org.apache.tomcat.util.http.Parameters   : Set encoding to UTF-8
2022-06-10 21:08:54.025 DEBUG 75288 --- [nio-8081-exec-3] o.s.web.servlet.DispatcherServlet        : GET "/swagger-ui.html", parameters={}
2022-06-10 21:08:54.033 DEBUG 75288 --- [nio-8081-exec-3] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped to ResourceHttpRequestHandler [classpath [META-INF/resources/]]
2022-06-10 21:08:54.037 DEBUG 75288 --- [nio-8081-exec-3] o.s.w.s.r.ResourceHttpRequestHandler     : Resource not found
2022-06-10 21:08:54.038 DEBUG 75288 --- [nio-8081-exec-3] o.s.web.servlet.DispatcherServlet        : Completed 404 NOT_FOUND
2022-06-10 21:08:54.040 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.c.C.[Tomcat].[localhost]           : Processing ErrorPage[errorCode=0, location=/error]
2022-06-10 21:08:54.041 DEBUG 75288 --- [nio-8081-exec-3] o.s.web.servlet.DispatcherServlet        : "ERROR" dispatch for GET "/error", parameters={}
2022-06-10 21:08:54.043 DEBUG 75288 --- [nio-8081-exec-3] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController#errorHtml(HttpServletRequest, HttpServletResponse)
2022-06-10 21:08:54.051 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       : getResource(META-INF/resources/error/404.html)
2022-06-10 21:08:54.052 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   Delegating to parent classloader jdk.internal.loader.ClassLoaders$AppClassLoader@5c29bfd
2022-06-10 21:08:54.055 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   --> Resource not found, returning null
2022-06-10 21:08:54.055 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       : getResource(resources/error/404.html)
2022-06-10 21:08:54.055 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   Delegating to parent classloader jdk.internal.loader.ClassLoaders$AppClassLoader@5c29bfd
2022-06-10 21:08:54.057 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   --> Resource not found, returning null
2022-06-10 21:08:54.057 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       : getResource(static/error/404.html)
2022-06-10 21:08:54.059 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   Delegating to parent classloader jdk.internal.loader.ClassLoaders$AppClassLoader@5c29bfd
2022-06-10 21:08:54.060 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   --> Resource not found, returning null
2022-06-10 21:08:54.060 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       : getResource(public/error/404.html)
2022-06-10 21:08:54.060 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   Delegating to parent classloader jdk.internal.loader.ClassLoaders$AppClassLoader@5c29bfd
2022-06-10 21:08:54.061 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   --> Resource not found, returning null
2022-06-10 21:08:54.062 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       : getResource(META-INF/resources/error/4xx.html)
2022-06-10 21:08:54.062 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   Delegating to parent classloader jdk.internal.loader.ClassLoaders$AppClassLoader@5c29bfd
2022-06-10 21:08:54.063 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   --> Resource not found, returning null
2022-06-10 21:08:54.063 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       : getResource(resources/error/4xx.html)
2022-06-10 21:08:54.064 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   Delegating to parent classloader jdk.internal.loader.ClassLoaders$AppClassLoader@5c29bfd
2022-06-10 21:08:54.067 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   --> Resource not found, returning null
2022-06-10 21:08:54.070 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       : getResource(static/error/4xx.html)
2022-06-10 21:08:54.070 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   Delegating to parent classloader jdk.internal.loader.ClassLoaders$AppClassLoader@5c29bfd
2022-06-10 21:08:54.071 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   --> Resource not found, returning null
2022-06-10 21:08:54.072 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       : getResource(public/error/4xx.html)
2022-06-10 21:08:54.072 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   Delegating to parent classloader jdk.internal.loader.ClassLoaders$AppClassLoader@5c29bfd
2022-06-10 21:08:54.073 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.loader.WebappClassLoaderBase       :   --> Resource not found, returning null
2022-06-10 21:08:54.074 DEBUG 75288 --- [nio-8081-exec-3] o.s.web.servlet.DispatcherServlet        : Exiting from "ERROR" dispatch, status 404
2022-06-10 21:08:54.075 DEBUG 75288 --- [nio-8081-exec-3] o.a.c.c.C.[.[.[/].[dispatcherServlet]    :  Disabling the response for further output
2022-06-10 21:08:54.076 DEBUG 75288 --- [nio-8081-exec-3] o.a.coyote.http11.Http11InputBuffer      : Before fill(): parsingHeader: [true], parsingRequestLine: [true], parsingRequestLinePhase: [0], parsingRequestLineStart: [0], byteBuffer.position(): [0], byteBuffer.limit(): [0], end: [935]
2022-06-10 21:08:54.076 DEBUG 75288 --- [nio-8081-exec-3] o.a.tomcat.util.net.SocketWrapperBase    : Socket: [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]], Read from buffer: [0]
2022-06-10 21:08:54.076 DEBUG 75288 --- [nio-8081-exec-3] org.apache.tomcat.util.net.NioEndpoint   : Socket: [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]], Read direct from socket: [0]
2022-06-10 21:08:54.077 DEBUG 75288 --- [nio-8081-exec-3] o.a.coyote.http11.Http11InputBuffer      : Received []
2022-06-10 21:08:54.077 DEBUG 75288 --- [nio-8081-exec-3] o.apache.coyote.http11.Http11Processor   : Socket: [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]], Status in: [OPEN_READ], State out: [OPEN]
2022-06-10 21:08:54.080 DEBUG 75288 --- [nio-8081-exec-3] org.apache.tomcat.util.net.NioEndpoint   : Registered read interest for [org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper@49e76d18:org.apache.tomcat.util.net.NioChannel@3a6d0e08:java.nio.channels.SocketChannel[connected local=/0:0:0:0:0:0:0:1:8081 remote=/0:0:0:0:0:0:0:1:51305]]
2022-06-10 21:09:21.840 DEBUG 75288 --- [alina-utility-2] org.apache.catalina.session.ManagerBase  : Start expire sessions StandardManager at 1654888161830 sessioncount 0
2022-06-10 21:09:21.840 DEBUG 75288 --- [alina-utility-2] org.apache.catalina.session.ManagerBase  : End expire sessions StandardManager processingTime 10 expired sessions: 0
```


Thanks

al-niessner commented:
@tloubrieu-jpl 

Yes. It is the upgrade to latest spring that did this. Not sure how to fix but needs another ticket as it is not related to this work. Although it might be some weird resource that I am not aware of that needs updating because of name changes in this work. However when googled, it seems to be known and related to latest version of spring.

See #141

tloubrieu-jpl commented:
Thanks @al-niessner  for creating the ticket

tloubrieu-jpl commented:
Hi @al-niessner ,

Weirdly, I am having this error again that we were not able to reproduce in ticket #133 

The request:
```
curl --location --request GET 'http://localhost:8081/uid/urn:nasa:pds:insight_rad:data_derived:hp3_rad_der_00014_20181211_073042::1.0/referencing/bundle' \
--header 'Accept: application/json'
```

Returns again:
`{"summary":{"q":"","hits":0,"took":16,"start":0,"limit":100,"sort":[],"properties":[]},"data":[]}%  `

The same thing happen with this request:
```
curl --location --request GET 'http://localhost:8081/uid/urn:nasa:pds:insight_rad:data_derived::7.0/referencing/bundle' \
--header 'Accept: application/json'
```

al-niessner commented:
@tloubrieu-jpl 

My Python unit tests stay that those do work in the new code. Does not verify it is the right lidvid but verifies that it 1 returned and that an actual lidvid comes with it.

Being that you thought release 1.0.1 had same problem I am going to say it is your test harness and/or database. Are you testing by hand or are you using postman?

Commit:
updates in prep for future work

Strange JUnit thing where org.junit.Test does not work but org.junit.jupiter.Test does. Seems the antlr unit tests are back on line.

Tried updating to todays latest spring to see if that fixes the root white label error problem.

Commit:
fixes issue 141

It is swagger 2 that change the help. It is now /swagger-ui/index.html. All references have been updated.

al-niessner commented on model/pom.xml:
```diff
@@ -119,20 +119,25 @@
-    </dependency>
-    
+    </dependency-->
+    <dependency>
```
@sonatype-lift ignore

al-niessner commented on service/pom.xml:
```diff
@@ -129,7 +129,12 @@ POSSIBILITY OF SUCH DAMAGE.
+                </dependency>
 
 		<!-- https://mvnrepository.com/artifact/org.springframework/spring-web -->
 		<dependency>
```
@sonatype-lift ignore

al-niessner commented on service/pom.xml:
```diff
@@ -129,7 +129,12 @@ POSSIBILITY OF SUCH DAMAGE.
 			<version>${springfox-version}</version>
-		</dependency>
+		</dependency-->
+                <dependency>
```
@sonatype-lift ignore

Commit:
remove unused

Commit:
remove unused

tloubrieu-jpl commented:
Dev review:

- [x] https://github.com/NASA-PDS/registry-api/issues/141, minor issue created https://github.com/NASA-PDS/registry-api/issues/149
- [x] https://github.com/NASA-PDS/registry-api/issues/133 , new ticket created https://github.com/NASA-PDS/registry-api/issues/150
- [x] https://github.com/NASA-PDS/registry-api/issues/131
- [x] https://github.com/NASA-PDS/registry-api/issues/128 , it does not sound like this ticket has been solved in this pull request, I will remove it from the list of fixes
- [x] https://github.com/NASA-PDS/registry-api/issues/118
- [x] https://github.com/NASA-PDS/registry-api/issues/12
- [x] https://github.com/NASA-PDS/registry-api/issues/11

tloubrieu-jpl reviewed (approved):
Thanks @al-niessner I successfully tested all the tickets embedded in this pull request.

I only removed #128 which does not sound like it is part of the PR (it is closed anyway).

I created 2 minor tickets:
#150  which is alike #133, hopefully you will be able to reproduce this time. I need you to help me to understand why I am having this behavior.
#149 a minor ticket on the newly repaired swagger-ui

Thanks again, you can merge this PR

Thomas
